### PR TITLE
Unify wrappers for the Ably SDK

### PR DESCRIPTION
**Pull request by lukasz-szyszkowski**
The main purpose of this PR is to create the universal wrapper for `ARTRealtime` from the `ably-cocoa` SDK and use the same either in the `subscriber` or `publisher` package. 

It's split into two parts:

1. Creating protocols for `publisher`, `subscriber` and `common`. Creating wrapper around the `ARTRealtime` using protocols. - this part doesn't contain the implementation of the "new wrapper" in the `DefaultPublisher` and `DefaultSubscriber` (commits: fb521295c43dd27c4e02abe02b116ccc7d6ec5de f291d04dd54f771a05b90d1eb6e8cdae27a4cfef 984abfba7239f8518426edbe895412855f25abf1)
2. Implementing of the before created wrapper in `DeafaultPublisher` and `DefaultSubscriber` using their protocols to hide some of the methods from each other. Fixing existing tests. 
   - [Publisher](https://github.com/ably/ably-asset-tracking-swift/pull/259#issuecomment-1034862583)
   - [Subscriber](https://github.com/ably/ably-asset-tracking-swift/pull/259#issuecomment-1036152752)

### CODE REVIEW

**Review by KacperKluka — APPROVED** [review]
Looks good (given my little Swift and iOS knowledge) :wink:

**Comment by maratal** [comment]
Would be nice to see an example of code duplication this PR solves. Thanks @lukasz-szyszkowski  🙏

**Comment by lukasz-szyszkowski** [comment]
> Would be nice to see an example of code duplication this PR solves. Thanks @lukasz-szyszkowski 🙏

Initialisers are almost the same. It's split into two methods [init](https://github.com/ably/ably-asset-tracking-swift/pull/259/commits/984abfba7239f8518426edbe895412855f25abf1#diff-62e19719dafe7413f0df947a41c3ac9e387137918e6f6e340dfa5721cc9db590R17-R21) and  [connect](https://github.com/ably/ably-asset-tracking-swift/pull/259/commits/984abfba7239f8518426edbe895412855f25abf1#diff-62e19719dafe7413f0df947a41c3ac9e387137918e6f6e340dfa5721cc9db590R23-R28)
---
CleanShot 2022-02-10 at 08 31 31@2x

They implement the same subscriptions like `channel.presence.subscribe`, `channel.presence.enter`, `channel.on`, `client.connection.on`
---
CleanShot 2022-02-10 at 08 48 08@2x
CleanShot 2022-02-10 at 08 48 55@2x
CleanShot 2022-02-10 at 08 51 29@2x
CleanShot 2022-02-10 at 08 56 49@2x

The `stopTracking` and `leaveChannelPresence` methods do the same
---
CleanShot 2022-02-10 at 08 59 18@2x

**Comment by lukasz-szyszkowski** [comment]
Here is another part of the changes related to `AblyPublisher` and its tests:
- Replacing `DefaultAblyPublisherService` with `DefaultAbly` wrapper hidden behind `AblyPublisher` interface (78113916ef8996d7f993a58d48694c1a26c5d75b)
- Updating `MockAblyPublisherService` from test target (fff3af466ccab76000899a14d9cf2bbacf38cde3)
- Updating `SystemTests` with new Ably wrapper (aa795559803a381522b948bf14f2b9dde05005ca)
- Fixing `Publisher` tests against new Ably wrapper (b89fed1d0c52ddfefa976bacdfff7b82072e7b9f)

**Review by KacperKluka — COMMENTED** [review]
Looks good, just one question :wink:

File: Sources/AblyAssetTrackingPublisher/DefaultPublisher.swift [hunk 1]
```diff
@@ -63,7 +65,7 @@ class DefaultPublisher: Publisher {
          logConfiguration: LogConfiguration,
          routingProfile: RoutingProfile,
          resolutionPolicyFactory: ResolutionPolicyFactory,
-         ablyService: AblyPublisherService,
+         ablyService: AblyPublisher,
```
[the remark this answers is not in the document]
Comment (lukasz-szyszkowski) [reply Sources/AblyAssetTrackingPublisher/DefaultPublisher.swift]: here 4d1547429648454a024809bad1af1be0eb922d3f and here b95137b6ace95d00b0ecf43cf3b2143ad63df887

**Review by KacperKluka — APPROVED** [review]

**Comment by lukasz-szyszkowski** [comment]
Yet another part of the changes related to `AblySubscriber` and its tests:
- Replacing `DefaultAblySubscriberService` with `DefaultAbly` wrapper hidden behind `AblySubscriber` interface (ca8379fcafa0fd9dc49868a484a69255785318ca)
- Updating `MockAblySubscriberService` from test target (f6c72e14f988689c380325ebafd9eff1fe4e6fed)
- Fixing `Subscriber` tests against new Ably wrapper (118cc36297ed07f57d83ab6aa5a0bfd1648bcee6)
- Remove previous protocols and wrappers for `Ably` (6d5e2d386b4654df6e1f761bffca486bc071795a)
- Add missing `subscriber` delegate method call (af574a51ff731253d43b763e4bced5515b3a8ccd)

**Review by KacperKluka — APPROVED** [review]

File: Sources/AblyAssetTrackingInternal/AblyWrapper/AblySubscriber.swift [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+import CoreLocation
+
+public protocol AblySubscriberServiceDelegate: AnyObject {
+    /**
+     Tells the delegate that `Ably` client connection state changed.
+     
+     - Parameter sender:    The `AblySubscriber` object which is delegating the change.
+     - Parameter state:     The `ConnectionState` object
+     */
+    func subscriberService(sender: AblySubscriber, didChangeClientConnectionStatus status: ConnectionState)
```
Comment (maratal) [inline Sources/AblyAssetTrackingInternal/AblyWrapper/AblySubscriber.swift:12]: didChangeClientConnectionStatus: status vs state, I think they should be called identically in both delegates

File: Sources/AblyAssetTrackingInternal/AblyWrapper/DefaultAbly.swift [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+    public func disconnect(trackableId: String, presenceData: PresenceData, completion: @escaping ResultHandler<Bool>) {
+        guard let channelToRemove = channels[trackableId] else {
+            completion(.success(false))
+            
+            return
+        }
+        
+        // Force cast intentional here. It's a fatal error if we are unable to create presenceData JSON
+        //
+        channelToRemove.presence.leave(try! presenceData.toJSONString()) { [weak self] error in
```
Comment (maratal) [inline Sources/AblyAssetTrackingInternal/AblyWrapper/DefaultAbly.swift:76]: There are not gonna be any useful crash message here, just "nil on force unwrap". Maybe explicitly add `precondition` crash in these cases? @lawrence-forooghian  WDYT

Comment (lukasz-szyszkowski) [reply Sources/AblyAssetTrackingInternal/AblyWrapper/DefaultAbly.swift:76]: Agree, I'll change it

File: Sources/AblyAssetTrackingInternal/AblyWrapper/AblySubscriber.swift (hunk 2 above)
Comment (lukasz-szyszkowski) [reply Sources/AblyAssetTrackingInternal/AblyWrapper/AblySubscriber.swift:12]: yeah, this is a copy-paste from previous delegation implementation (I didn't touch this one), but I agree -> I'll change it for the coherence

File: Sources/AblyAssetTrackingInternal/AblyWrapper/DefaultAbly.swift (hunk 3 above)
Comment (lukasz-szyszkowski) [reply Sources/AblyAssetTrackingInternal/AblyWrapper/DefaultAbly.swift]: 1fa230ac04856cfa97143ea13e3158d5961b825c

**Review by maratal — APPROVED** [review]

**Pull request merged**